CORE-13863 - extra crypto metrics

conal-smith opened the pull request:
Helm chart changes - https://github.com/corda/engineering-helm/pull/37

conal-smith committed:
CORE-13863 - extra crypto metrics

conal-smith committed:
CORE-13863 - a few minor fixes

conal-smith committed:
CORE-13863 - fine tune some crypto metrics

conal-smith committed:
CORE-13863 - fix osgi error

JamesHR3 commented on components/crypto/crypto-persistence/src/main/kotlin/net/corda/crypto/persistence/GetEntityManagerFactory.kt:
```diff
@@ -17,7 +17,7 @@ fun getEntityManagerFactory(
 ): EntityManagerFactory {
     return CordaMetrics.Metric.EntityManagerFactoryCreationTimer.builder()
-        .withTag(CordaMetrics.Tag.VirtualNode, tenantId)
+        .withTag(CordaMetrics.Tag.Tenant, tenantId)
```
Is there a distinction between tenant and virtual node?

kyriathar replied:
So I believe tenant can be virtual node (id) but also "cluster level tenants" like `CRYPTO`, `P2P` and other categories under which keys are added.

sprintengine replied:
Yeah tenant is either a vnodeId or "CRYPTO", "P2P", "REST"

JamesHR3 commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt:
```diff
@@ -69,10 +70,22 @@ class CryptoServiceFactoryImpl @Activate constructor(
 
     override fun getInstance(hsmId: String): CryptoService =
-        impl.getInstance(hsmId)
+        CordaMetrics.Metric.CryptoServiceInstanceCreationTimer.builder()
```
Any reason for not using the helper you have defined further down for `GetInstance` here?

sprintengine replied:
Hmm it doesn't really feel correct to have a dependency on softhsm-impl for a metric that records getInstance.

Maybe I could find a better home for such a metric, crypto-core perhaps, though not sure I want to add metrics dependency in there.

sprintengine replied:
Removed this helper anyway, following below changes

JamesHR3 commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt:
```diff
@@ -69,10 +70,22 @@ class CryptoServiceFactoryImpl @Activate constructor(
 
     override fun findInstance(tenantId: String, category: String): CryptoServiceRef =
-        impl.findInstance(tenantId, category)
+        CordaMetrics.Metric.CryptoServiceInstanceCreationTimer.builder()
```
Should this be the same metric or its own one? Is a new instance being created?

sprintengine replied:
I've split out some of the instance creation timers:
- CryptoServiceFindInstanceTimer with label on instance.type ("SoftCryptoServiceImpl")
- CryptoServiceGetInstanceTimer with label on instance.type ("SoftCryptoServiceImpl")
- CryptoSigningRepositoryGetInstanceTimer with label on tenant and instance.type (maybe unnecessary)
- I removed recordGetInstance from softCryptoServiceProvider because this should be wrapped by the `CryptoServiceGetInstanceTimer` already.

JamesHR3 commented on libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt:
```diff
@@ -368,6 +368,16 @@ object CordaMetrics {
+        /**
+         * The time taken to create entity manager factories.
+         */
+        object CryptoMethodTimer: Metric<Timer>("crypto.method.time", CordaMetrics::timer)
```
One of these kdocs is incorrect.

I have a separate concern that by putting a generic method timer in here it's going to be very hard to determine what exactly is being timed from a user perspective. Should we restrict this to a particular service and tag it by operation for that service?

sprintengine replied:
Yeah I was a bit iffy on this metric, I've split out some of these metrics for example:
- GetOwnedKeyRecordTimer (label on operationName = "deriveSharedSecret" / "sign")
- CryptoCipherSchemeTimer (label on operationName = "decodePublicKeyFromByteArray" / "decodePublicKeyFromString" / "encodePublicKeyToString")
- CryptoSignatureSpecTimer (label on operationName = "toSignatureSpec" / "toWire")

Hopefully this should be a bit better

charlieR3 commented on libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt:
```diff
@@ -571,8 +581,32 @@ object CordaMetrics {
+        /**
+         * Label to identify the method inside a class / implementation.
+         */
+        Method("method"),
```
Could you use `OperationName` instead of adding another tag?

sprintengine replied:
Yeah Method is gone, cheers

charlieR3 commented on libs/metrics/src/main/kotlin/net/corda/metrics/CordaMetrics.kt:
```diff
@@ -571,8 +581,32 @@ object CordaMetrics {
+         */
+        Method("method"),
+
+        /**
+         * Label to identify the method inside a class / implementation.
+         */
+        PublicKeyType("publickey.type"),
+
+        /**
+         * Label to identify the method inside a class / implementation.
+         */
+        Tenant("tenant"),
```
kdocs are wrong

sprintengine replied:
Thanks yeah I didn't update quite a few, they should all be updated now

charlieR3 commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt:
```diff
@@ -374,6 +377,11 @@ class SigningServiceImpl(
         if (signingKeyInfo.fullId != requestedFullKeyId) throw IllegalArgumentException(
             "The tenant $tenantId doesn't own public key '${publicKey.publicKeyId()}'."
         )
+        CordaMetrics.Metric.CryptoMethodTimer.builder()
```
Could you wrap the whole function in `recordCallable` instead of calculating the duration yourself?

sprintengine replied:
Slightly changed this around so I've now a "GetOwnedKeyRecordTimer" with label on whether is from "sign" or "deriveSharedSecret" operation.

charlieR3 commented on libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt:
```diff
@@ -77,6 +82,11 @@ fun CryptoSignatureSpec.toSignatureSpec(serializer: AlgorithmParameterSpecEncodi
+        CordaMetrics.Metric.CryptoMethodTimer.builder()
+            .withTag(CordaMetrics.Tag.Method, CRYPTO_SIGNATURE_SPEC_TO_SIGNATURE_SPEC_METHOD_NAME)
+            .build()
+            .record(Duration.between(startTime, Instant.now()))
```
Could this be a recordCallable also?

sprintengine replied:
Slightly changed this around so I've now a "CryptoSignatureSpecTimer" with label on operationName (toSignatureSpec or toWire).

I left it in the .also{} at the end because I think it is slightly more readable and I don't need the non-null !! casting.

conal-smith committed:
CORE-13863 - review comments:
- fix kdocs for the metrics
- get rid of method based metric and replace with finer-grained metric names using operationName as a tag
- removed metricUtils, separate get instance metrics for each provider/factory domain

kyriathar commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt:
```diff
@@ -1,5 +1,7 @@
 package net.corda.crypto.service.impl
 
+import java.time.Duration
+import java.time.Instant
```
These imports need be re-ordered I think

kyriathar commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt:
```diff
@@ -68,11 +70,25 @@ class CryptoServiceFactoryImpl @Activate constructor(
-        impl.getInstance(hsmId)
+    override fun getInstance(hsmId: String): CryptoService {
+        val startTime = Instant.now()
+        return impl.getInstance(hsmId).also {
```
Not sure about this one. We are adding a metric around something that most of the time will be there in place (`CryptoServiceFactoryImpl.Impl.cryptoService`) or maybe I am missing something (other than the first time which will be initialization through `lazy`). Maybe @dickon can help us here.

sprintengine replied:
Think this one gets called from sign(), and calls the lazy initializer on cryptoService

kyriathar commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt:
```diff
@@ -68,11 +70,25 @@ class CryptoServiceFactoryImpl @Activate constructor(
-        impl.findInstance(tenantId, category)
+    override fun findInstance(tenantId: String, category: String): CryptoServiceRef {
+        val startTime = Instant.now()
+        return impl.findInstance(tenantId, category).also {
```
Same with `getInstance` I think although this one also involves a DB query, so maybe makes more sense to have a metric. Although I don't think it gets involved in flow-crypto operations.

sprintengine replied:
Ok I can remove this findInstance if it isn't relevant. Along with deriveSharedKey call to getOwnedRecords

kyriathar commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt:
```diff
@@ -277,7 +286,15 @@ class SigningServiceImpl(
-        val record = getOwnedKeyRecord(tenantId, publicKey)
+        val record =
+            CordaMetrics.Metric.GetOwnedKeyRecordTimer.builder()
+                .withTag(CordaMetrics.Tag.OperationName, DERIVE_SHARED_SECRET_OPERATION_NAME)
```
I don't think this operation takes place in flow-crypto interactions either.

sprintengine replied:
So `deriveSharedSecret` won't be called during flows? I think for now I might leave this in, and later we can remove a few of these if they prove to be not useful.

kyriathar commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SigningRepositoryFactoryImpl.kt:
```diff
@@ -18,16 +19,23 @@ class SigningRepositoryFactoryImpl(
+            .build()
+            .recordCallable {
+                SigningRepositoryImpl(
+                    entityManagerFactory = getEntityManagerFactory(
```
I think this metric could be a metric around `getEntityManagerFactory`. Rest of it seems static therefore maybe metric noise?

sprintengine replied:
GetEntityManagerFactory is called in a few places, so at least this gives us detail on time for signing repository.

kyriathar commented on libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt:
```diff
@@ -140,31 +145,48 @@ class CipherSchemeMetadataProvider : KeyEncodingService {
-        throw e
-    } catch (e: Throwable) {
-        throw CryptoException("Failed to decode public key", e)
+    private fun <T : Any> recordCipherOperation(operationName: String, op: () -> T): T {
```
I don't think there is any cipher operation going on here, is more like serializing/ deserializing keys I think.

sprintengine replied:
Thanks I renamed this to something slightly better

kyriathar commented on libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt:
```diff
@@ -77,23 +83,38 @@ fun CryptoSignatureSpec.toSignatureSpec(serializer: AlgorithmParameterSpecEncodi
         else -> SignatureSpecImpl(signatureName)
+    }.also {
+        CordaMetrics.Metric.CryptoSignatureSpecTimer.builder()
+            .withTag(CordaMetrics.Tag.OperationName, TO_SIGNATURE_SPEC_OPERATION_NAME)
```
Not really sure this metric adds much, I believe it's just Avro signature spec to DTO.

sprintengine replied:
It should include some deserializing too

kyriathar commented on libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/WireUtils.kt:
```diff
@@ -77,23 +83,38 @@ fun CryptoSignatureSpec.toSignatureSpec(serializer: AlgorithmParameterSpecEncodi
         else -> CryptoSignatureSpec(signatureName, null, null)
+    }.also {
+        CordaMetrics.Metric.CryptoSignatureSpecTimer.builder()
+            .withTag(CordaMetrics.Tag.OperationName, TO_WIRE_OPERATION_NAME)
```
Same as above, DTO to Avro, seems quite static in terms of time?

sprintengine replied:
Should include some serializing too

kyriathar reviewed (commented):
Some minor comments, otherwise `sign` and `filterMyKeys` paths are covered.

conal-smith committed:
CORE-13863 - rename method, remove instance type as it is not useful

conal-smith committed:
CORE-13863 - put crypto worker metrics into their own object

conal-smith committed:
CORE-13863 - a few excess metric removals

kyriathar commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/CryptoServiceFactoryImpl.kt:
```diff
@@ -71,8 +71,9 @@ class CryptoServiceFactoryImpl @Activate constructor(
 
-    override fun getInstance(hsmId: String): CryptoService =
-        impl.getInstance(hsmId)
+    override fun getInstance(hsmId: String): CryptoService {
+        return impl.getInstance(hsmId)
+    }
```
Can we please completely revert this change (to keep original git message).

kyriathar commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt:
```diff
@@ -3,6 +3,7 @@ package net.corda.crypto.service.impl
 import com.github.benmanes.caffeine.cache.Cache
 import com.github.benmanes.caffeine.cache.Caffeine
 import java.security.InvalidParameterException
+import java.security.PublicKey
 import java.util.concurrent.TimeUnit
```
I think this could do with some imports optimizing

kyriathar commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoService.kt:
```diff
@@ -199,33 +200,34 @@ class SoftCryptoService(
+                signature.sign()
+            }
+        }
+        CordaMetrics.Metric.Crypto.SoftSignTimer
```
Perhaps just call `SignTimer`? We don't need "Soft" I think it's confusing (We don't support Hardware HSMs so word "Soft" should just be removed completely at some point).

kyriathar commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoService.kt:
```diff
@@ -199,33 +200,34 @@ class SoftCryptoService(
+        }
+        CordaMetrics.Metric.Crypto.SoftSignTimer
             .builder()
+            .withTag(CordaMetrics.Tag.AlgorithmName, spec.keyScheme.algorithmName)
```
I think we should rather tag `SigningSpec.signatureSpec` here and not just key's algorithm.

kyriathar commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoServiceProviderImpl.kt:
```diff
@@ -5,6 +5,13 @@ import com.github.benmanes.caffeine.cache.Caffeine
 import com.typesafe.config.ConfigList
 import com.typesafe.config.ConfigObject
 import com.typesafe.config.ConfigValue
+import java.security.InvalidParameterException
+import java.security.KeyPairGenerator
+import java.security.PrivateKey
+import java.security.Provider
+import java.security.PublicKey
+import java.util.InvalidPropertiesFormatException
+import java.util.concurrent.TimeUnit
```
imports should be re-ordered here

sprintengine replied:
These appear to be correct according to my IJ settings, what did your optimize imports change?

kyriathar commented on libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt:
```diff
@@ -1,9 +1,16 @@
 package net.corda.crypto.impl
 
+import java.io.StringReader
+import java.io.StringWriter
+import java.security.Provider
+import java.security.PublicKey
+import java.security.SecureRandom
+import java.security.spec.X509EncodedKeySpec
```
imports

conalsmith-r3 replied:
Both of these are correct according to my IJ, IJ is a bit inconsistent on the order of these across machines.

sprintengine replied:
Same here, when you optimize imports what changes here?

kyriathar reviewed (commented):
A few last minor comments

conal-smith committed:
CORE-13863 - rename sign timer, change tag to signature spec, fix some imports

kyriathar reviewed (approved):
LGTM
